org.eclipse.tm4e.core/src/main/java/org/eclipse/tm4e/core/model/AbstractTMModel.java

```diff
@@ -128,170 +70,203 @@ public int getSize() {
 		 * @param name
 		 *            the thread's name
 		 */
-		public BackgroundThread(String name) {
+		public TokenizerThread(String name, AbstractTMModel model) {
 			super(name);
+			this.model = model;
+			this.tokenizer = new Tokenizer(model.getGrammar());
 			setPriority(Thread.MIN_PRIORITY);
 			setDaemon(true);
 		}
 
-		/**
-		 * Suspends the caller of this method until this background thread has
-		 * emptied the dirty region queue.
-		 */
-		public void suspendCallerWhileDirty() {
-			boolean isDirty;
-			do {
-				synchronized (fDirtyRegionQueue) {
-					isDirty = fDirtyRegionQueue.getSize() > 0;
-					if (isDirty) {
+		@Override
+		public void run() {
+			if (interrupted())
+				return;
+
+			// initialize
+			int linesLength = model.getNumberOfLines();
+			for (int line = 0; line < linesLength; line++) {
+				try {
+					model.lines.addLine(line);
+				} catch (Exception e) {
+					e.printStackTrace();
+				}
+			}
+			model.lines.forEach(ModelLine::resetTokenizationState);
+			model.lines.get(0).setState(tokenizer.getInitialState());
+			model._invalidLineStartIndex = 0;
+			
+			while (!interrupted()) {

```

Comment:
The interrupted state flag is supposed to cover the use-case, there shouldn't have need to add another flag that do that. Locally, on Linux, I can see the Thread running when editor is open (I see it when pausing on the process with the debugger) and the Thread not present any more when closing the editor (same thing, closed editor than paused process to look at threads).
Can you please try with `isInterrupted()` instead of `interrupted()` here?